libkbfs/bare_root_metadata_v2_test.go

```diff
@@ -265,3 +265,472 @@ func TestRevokeRemovedDevicesV2(t *testing.T) {
+			pubKey := privKey.GetPublicKey()
+			if pubKeys[uid] == nil {
+				pubKeys[uid] = make(map[kbfscrypto.CryptPublicKey]bool)
+			}
+			pubKeys[uid][pubKey] = true
+		}
+	}
+	return pubKeys
+}
+
+// expecteRekeyInfoV2 contains all the information needed to check a
+// rekey run (that doesn't add a generation).
+//
+// If both writerPrivKeys and readerPrivKeys are empty, then
+// ePubKeyIndex and ePubKey are ignored.
+type expecteRekeyInfoV2 struct {
+	writerPrivKeys, readerPrivKeys userDevicePrivateKeys
+	serverHalves                   UserDeviceKeyServerHalves
+	ePubKeyIndex                   int
+	ePubKey                        kbfscrypto.TLFEphemeralPublicKey
+}
+
+func checkCryptKeyInfo(t *testing.T, privKey kbfscrypto.CryptPrivateKey,
+	serverHalf kbfscrypto.TLFCryptKeyServerHalf, expectedEPubKeyIndex int,
+	expectedEPubKey kbfscrypto.TLFEphemeralPublicKey,
+	expectedTLFCryptKey kbfscrypto.TLFCryptKey, info TLFCryptKeyInfo,
+	ePubKey kbfscrypto.TLFEphemeralPublicKey) {
+	dummySigningKey := kbfscrypto.MakeFakeSigningKeyOrBust("dummy")
+	codec := kbfscodec.NewMsgpack()
+	crypto := NewCryptoLocal(codec, dummySigningKey, privKey)
+
+	require.Equal(t, expectedEPubKeyIndex, info.EPubKeyIndex)
+	require.Equal(t, expectedEPubKey, ePubKey)
+
+	ctx := context.Background()
+	clientHalf, err := crypto.DecryptTLFCryptKeyClientHalf(
+		ctx, ePubKey, info.ClientHalf)
+	require.NoError(t, err)
+
+	tlfCryptKey, err := crypto.UnmaskTLFCryptKey(serverHalf, clientHalf)
+	require.NoError(t, err)
+	require.Equal(t, expectedTLFCryptKey, tlfCryptKey)
+}
+
+// checkGetTLFCryptKeyV2 checks that wkb and rkb contain the info
+// necessary to get the TLF crypt key for each user in expected, which
+// must all match expectedTLFCryptKey.
+func checkGetTLFCryptKeyV2(t *testing.T, expected expecteRekeyInfoV2,
+	expectedTLFCryptKey kbfscrypto.TLFCryptKey,
+	wkb *TLFWriterKeyBundleV2, rkb *TLFReaderKeyBundleV2) {
+	for uid, privKeys := range expected.writerPrivKeys {
+		for privKey := range privKeys {
+			pubKey := privKey.GetPublicKey()
+			serverHalf, ok := expected.serverHalves[uid][pubKey]
+			require.True(t, ok, "writer uid=%s, key=%s",
+				uid, pubKey)
+
+			info, ok := wkb.WKeys[uid][pubKey.KID()]
+			require.True(t, ok)
+
+			ePubKey := wkb.TLFEphemeralPublicKeys[info.EPubKeyIndex]
+
+			checkCryptKeyInfo(t, privKey, serverHalf,
+				expected.ePubKeyIndex, expected.ePubKey,
+				expectedTLFCryptKey, info, ePubKey)
+		}
+	}
+
+	for uid, privKeys := range expected.readerPrivKeys {
+		for privKey := range privKeys {
+			pubKey := privKey.GetPublicKey()
+			serverHalf, ok := expected.serverHalves[uid][pubKey]
+			require.True(t, ok, "reader uid=%s, key=%s",
+				uid, pubKey)
+
+			info, ok := rkb.RKeys[uid][pubKey.KID()]
+			require.True(t, ok)
+
+			var ePubKey kbfscrypto.TLFEphemeralPublicKey
+			if info.EPubKeyIndex >= 0 {
+				ePubKey = wkb.TLFEphemeralPublicKeys[info.EPubKeyIndex]
+			} else {
+				ePubKey = rkb.TLFReaderEphemeralPublicKeys[-1-info.EPubKeyIndex]
```

Comment:
Done, as `getEphemeralPublicKeyInfoV2`. It's an ugly function, since I had to make upconversion use it too, but now there's only one instance of "-1 - something" in non-test code.